specific/cmo_sale_discount_total_enhance/models/sale.py

```diff
@@ -27,32 +31,42 @@ class sale_order(models.Model):
         track_visibility='always',
     )
 
-    @api.model
+    @api.multi
     def calculate_discount(self, vals):
+        self.ensure_one()
         discount = 0
+        last_line = False
         if ('discount_rate' in vals) and ('order_line' in vals):
             res = vals['sale_order_record']
+            cur = res.pricelist_id.currency_id
             if vals['discount_type'] == 'percent':
                 discount = vals['discount_rate']
+                order_lines = res.order_line
             else:
                 amount = sum(res.order_line.mapped(
                     lambda r: r.product_uom_qty * r.price_unit)
                 )
                 total = amount if amount != 0 else 1 # prevent devison by zero
                 discount_rate = vals['discount_rate']
-                if discount_rate != 0:
-                    discount = (discount_rate / total) * 100.0
-                else:
-                    discount = discount_rate
-            for line in res.order_line:
+                discount = round((discount_rate / total) * 100.0, 16)
+                order_lines = res.order_line[:-1]
+                last_line = res.order_line[-1]
+            discount_line = 0
+            for line in order_lines:
                 line.write({'discount': discount})
+                discount_line += cur.round(
+                    line.price_unit * line.product_uom_qty * discount / 100.0
+                    )
+            if last_line:
+                last_discount = (cur.round(discount_rate - discount_line) / \
+                    last_line.price_unit) * 100.0
+                last_line.write({'discount': last_discount})
 
     @api.model
     def create(self, vals):
         res = super(sale_order, self).create(vals)
         vals['sale_order_record'] = res
```

Comment:
This part of code looks doubtful. Are you passing an object (res) into vals ??? I think it is not necessary.
Actually, the "res" is already the new sale.order object. You can just call 
`res.calculate_discount()`
No need to pass anything. The **self** in the calculate_discount() already have all the value you will need. Isn't it?